":read :cmd" reads Ex command ":cmd" output

bwalshe opened the pull request:
Based on issue #30376, I have made this change so that you can run commands when using the `read` command, and have their output added to the current buffer.

The current implementation uses vim script to basically do a `:put=execute('{cmd}')`, and I am happy to re-implement this in C if that would be better, but I would like to get some feedback on the tests and expected behaviour first. 

close #30376
close #31909

justinmk commented:
definitely wouldn't use a temp file. likely the best way to implement this is to re-use the internals of `:help execute()` : https://github.com/neovim/neovim/blob/e049c6e4c08a141c94218672e770f86f91c27a11/src/nvim/eval/funcs.c#L1592-L1594

Adding tests is the most valuable part of this work, because it defines the "specification", so then we can improve the implementation later. A good place for tests would be `test/functional/ex_cmds/` . There is currently no `read_spec.lua` in there, but perhaps could use `test/functional/ex_cmds/write_spec.lua` as a starting point.

You can [run a specific test](https://github.com/neovim/neovim/blob/master/test/README.md#filter-by-file) like this:

    TEST_FILE=test/functional/example_spec.lua make functionaltest

bwalshe commented:
OK thanks for the advice. It looks like I won't have time to work on this until some time next week, but I will probably have more questions on Tuesday or Wednesday.

bwalshe commented:
@justinmk I have pushed some tests and implemented the basic solution of having the C code defer to vimscript. 

As you can see one of the unit tests is failing as vim is not handling quotations the way I expected it to. It looks like the behaviour is slightly different depending on which code path you go down. For example, `:execute('echo "line1\nline2"')` will echo two lines, but  `$put =execute('echo "line1\nline2"')`  results in 

```
E115: Missing quote: 'echo
E116: Invalid arguments for function execute
```

Does that seem right to you?

Also what is the expected behaviour when interacting with `:split`? Right now if I do `:split +read !ls` it tells me that it cannot find the file `ls`. If split + read doesn't work with shell commands, should I also implement that?

bwalshe commented:
BTW - the way I am going, my commit history is going to be quite messy. Should I rebase as I go, or do it at the end?

zeertzjq commented:
> @justinmk I have pushed some tests and implemented the basic solution of having the C code defer to vimscript.
> 
> As you can see one of the unit tests is failing as vim is not handling quotations the way I expected it to. It looks like the behaviour is slightly different depending on which code path you go down. For example, `:execute('echo "line1\nline2"')` will echo two lines, but `$put =execute('echo "line1\nline2"')` results in
> 
> ```
> E115: Missing quote: 'echo
> E116: Invalid arguments for function execute
> ```
> 
> Does that seem right to you?
> 
> Also what is the expected behaviour when interacting with `:split`? Right now if I do `:split +read !ls` it tells me that it cannot find the file `ls`. If split + read doesn't work with shell commands, should I also implement that?

You cannot use `"` in a `:put` Ex command, as `"` starts a comment.

bwalshe commented on src/nvim/ex_docmd.c:
```diff
@@ -5640,6 +5641,20 @@ static void ex_syncbind(exarg_T *eap)
+  StringBuilder put_cmd = KV_INITIAL_VALUE;
+  kv_printf(put_cmd, "try | ");
+  kv_printf(put_cmd, "%dput=execute('%s') | ", eap->line2, cmd);
+  kv_printf(put_cmd, "execute 'norm! `.' | ");
```
`execute()` seems to add an extra empty line above the output, so I have to jump back to the top of the inserted text and delete this line, so that we get behaviour that is more similar to reading a file.

bwalshe commented on src/nvim/ex_docmd.c:
```diff
@@ -5640,6 +5641,20 @@ static void ex_syncbind(exarg_T *eap)
+{
+  char *cmd = eap->arg + 1;
+  StringBuilder put_cmd = KV_INITIAL_VALUE;
+  kv_printf(put_cmd, "try | ");
```
wrapping things in a `try` means that if there is an error with the command no text is inserted into the buffer. 

This _feels_ like the right behaviour to me, but potentially could someone want the error text inserted? When using `read !cmd` text will be inserted into the buffer even if there is an error.

bwalshe commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,93 @@
+  end)
+  it('inserts text from shell', function()
+    test_read('read !cat ' .. tmp_file, inserted_middle)
+    eq({ 0, 3, 3, 0 }, fn.getpos('.'))
```
`:read {path}` and `read !{cmd}` leave the cursor in different positions. For my implementation of `:read :cmd` I leave the cursor in the same position as when reading a file.

bwalshe commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,93 @@
+    eq("Vim(read):E484: Can't open file asdfasdf", pcall_err(execute, ':read asdfasdf'))
+  end)
+  it('errors out correctly when an invalid command is used', function()
+    eq('Vim:E492: Not an editor command: asdfasdf', pcall_err(execute, ':read :asdfasdf'))
```
I'm not sure if this error message is the best it could be. Should I have it say `Vim(read):E492: ...`

Alternatively, the error text could be written to the buffer. I don't _think_ it should, but I can see how someone might want that behaviour.

justinmk replied:
> Should I have it say `Vim(read):E492: ...`

if it's not much trouble then I think so.

> Alternatively, the error text could be written to the buffer

`execute()` appears to capture error messages. But for now, let's optimize for implementation simplicity. Can always revisit later if needed.

bwalshe replied:
It would actually be marginally simpler to have the error text written to the buffer. That is what would happen if I didn't wrap the vimscript up in a `try` clause. 

When you say we can revisit later on, do you mean before completing the pull request, or some time after this is merged? I imagine this would become a breaking change, so I would want to get it right.

justinmk replied:
I mean after this is merged. That's a good instinct and a good general practice, but sometimes features like this need to see real-world usage before we fully decide the interface. Especially when "the stakes are low", as with this particular behavior decision (error capture vs no error capture).

Errors are not the common case so I think it's better to just optimize for code simplicity here. Thuse removing the try/catch sounds good.

bwalshe replied:
Cool. I know it is pretty low stakes, and the best course of action is to see how it goes, but I didn't want to make an assumption.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5640,6 +5641,20 @@ static void ex_syncbind(exarg_T *eap)
+  char *cmd = eap->arg + 1;
+  StringBuilder put_cmd = KV_INITIAL_VALUE;
+  kv_printf(put_cmd, "try | ");
+  kv_printf(put_cmd, "%dput=execute('%s') | ", eap->line2, cmd);
```
will need to escape single-quotes, possibly other stuff. actually, all that could be avoided, including support for double quotes ("), by storing `cmd` in a temporary `g:_blah` var, then doing:

```suggestion
  kv_printf(put_cmd, "%dput=execute(g:_blah) | ", eap->line2, cmd);
```

bwalshe replied:
OK, I am now using an api call to put {cmd} directly into a global variable. I haven't updated the unit tests yet though as I think I am going to have to take some time to understand how things are being parsed before being to `ex_read`. For example, the big issue with double quotes (as I understand it right now) is that these cause the text to be marked as a comment and chopped off before `eap` is passed to `ex_read`.

justinmk replied:
yeah it's possible that double quotes can't easily be supported because the cmdline is parsed before control reaches the code you have added here. can forget about that for now.

zeertzjq commented on runtime/doc/insert.txt:
```diff
@@ -2000,14 +2000,20 @@ NOTE: These commands cannot be used with |:global| or |:vglobal|.
+
+							*:r:* *:read:*
+:[range]r[ead] :{cmd}
+			Execute vim {cmd} and insert its output below the
```
```suggestion
			Execute Ex command {cmd} and insert its output below the
```

zeertzjq commented on runtime/doc/insert.txt:
```diff
@@ -2000,14 +2000,20 @@ NOTE: These commands cannot be used with |:global| or |:vglobal|.
+
+							*:r:* *:read:*
+:[range]r[ead] :{cmd}
+			Execute vim {cmd} and insert its output below the
+			specified line
```
```suggestion
			specified line.
```

zeertzjq replied:
```suggestion
			Execute Ex command {cmd} and insert its output below
			the cursor or the specified line.
```

justinmk commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,93 @@
+  it('inserts text from command', function()
+    local make_lines = string.format('let lines="%s"', table.concat(read_text, '\\n'))
+    execute(make_lines)
+    test_read('read :echo lines', inserted_middle)
```
also need a test that checks behavior with `|`
```suggestion
    test_read('read :foo | bar | baz', inserted_middle)
```

justinmk replied:
still need this

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -5643,16 +5644,29 @@ static void ex_syncbind(exarg_T *eap)
+  kv_printf(put_cmd, "%dput=execute(g:%s) | ", eap->line2, cmd_var_name.data);
+  kv_printf(put_cmd, "execute 'norm! )`.' | ");
   kv_printf(put_cmd, "execute 'd _' | ");
   kv_printf(put_cmd, "endtry");
```
unlet the `g:` var after

justinmk replied:
still need this

justinmk commented on runtime/doc/news.txt:
```diff
@@ -159,6 +159,7 @@ EDITOR
   "~/" are now expanded to the user's profile directory, not a relative path
   to a literal "~" directory.
 • |hl-PmenuMatch| and |hl-PmenuMatchSel| show matched text in completion popup.
+• |:read:| can now execute an Ex command and insert the output into the buffer.
```
```suggestion
• |:read:| reads the output of an Ex command and into the buffer. Example:>
  :read :ls
```

bwalshe commented:
@justinmk Apologies for leaving this hanging so long. I had something come up at work. I'm going to rebase this and see how it looks.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -212,6 +212,8 @@ EDITOR
   "~/" are now expanded to the user's profile directory, not a relative path
   to a literal "~" directory.
 • |hl-PmenuMatch| and |hl-PmenuMatchSel| show matched text in completion popup.
+• |:read:| reads the output of an Ex command into the buffer. Example:> :read
+  :ls
```
```suggestion
• |:read:| reads the output of an Ex command into the buffer. Example:>
    :read
    :ls
<
```

justinmk commented on runtime/doc/insert.txt:
```diff
@@ -2010,14 +2010,20 @@ NOTE: These commands cannot be used with |:global| or |:vglobal|.
+							*:r:* *:read:*
+:[range]r[ead] :{cmd}
+			Execute Ex command {cmd} and insert its output below
+			the cursor or the specified line
```
```suggestion
			the cursor or the specified line.
```

justinmk commented on runtime/doc/insert.txt:
```diff
@@ -2010,14 +2010,20 @@ NOTE: These commands cannot be used with |:global| or |:vglobal|.
+			save the output of the command, which can be set to
+			include stderr or not.  {cmd} is executed like with
+			":!{cmd}", any '!' is replaced with the previous
+			command |:!|. See |++opt| for the possible values of
+			[++opt].
```
```suggestion
			command |:!|.
			See |++opt| for the possible values of [++opt].
```

justinmk commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,94 @@
+    write_file(tmp_file, table.concat(read_text, '\n'), true)
+  end)
+  after_each(cleanup)
+  it('inserts text from file', function()
+    test_read('read ' .. tmp_file, inserted_middle)
+    eq({ 0, 2, 2, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from shell', function()
+    test_read('read !cat ' .. tmp_file, inserted_middle)
```
these are thoroughly tested elsewhere. don't need them here.

justinmk commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,94 @@
+    execute(make_lines)
+    test_undo('read :echo lines')
+  end)
+  it('errors out correctly when a non-existant file is used', function()
+    eq("Vim(read):E484: Can't open file asdfasdf", pcall_err(execute, ':read asdfasdf'))
+  end)
+  it('errors out correctly when an invalid command is used', function()
```
just put these in a single `it('failure modes'` block.

justinmk commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,94 @@
+    eq('fileencoding=utf-8', vim.trim(execute('set fileencoding?')))
+    eq('nobomb', vim.trim(execute('set bomb?')))
+  end)
+  it('file reads can be undone', function()
+    test_undo('read ' .. tmp_file)
+  end)
+  it('shell reads can be undone', function()
+    test_undo('read !cat ' .. tmp_file)
```
drop these. we already have tests for this elsewhere.

justinmk commented on test/functional/ex_cmds/read_spec.lua:
```diff
@@ -0,0 +1,94 @@
+    execute(make_lines)
+    test_read('read :echo lines', inserted_middle)
+  end)
+  it('inserts text from file at specific position', function()
+    test_read('0read ' .. tmp_file, inserted_start)
+    eq({ 0, 1, 2, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from shell cmd at specific position', function()
```
drop

bwalshe committed:
feat(editor): :read reads the output of an Ex command into the buffer

Problem: :read can read from files or stdin, but not from other Ex commands

Solution: use `put=execute(cmd)` to execute cmd and send its output to the
buffer

bwalshe commented:
@justinmk I'm getting consistent failures on `Test_search_stat_option`/`macos intel clang`, but it it works fine on the other platforms and it is not immediately clear to me what is going on there.

bwalshe commented:
Hi, I am glad this is back up and in progress again, and I don't want to interfere with the work, but could I ask why we couldn't get this done originally? Is there something I could have done to move things along?

mikeslattery commented:
> Happy to create a new PR referencing this one if that's better, or wait for @mikeslattery to respond first.

@benarcher2691, It's not important, but I don't remember being part of this effort.  Perhaps you confused me with my activity in another project?
